src/elliptic/curves/traits.rs

```diff
@@ -5,48 +5,244 @@
     License MIT: <https://github.com/KZen-networks/curv/blob/master/LICENSE>
 */
 
-use std::ops::{Add, Mul};
+use std::fmt;
+
+use serde::{Deserialize, Serialize};
+use zeroize::Zeroize;
 
 use crate::BigInt;
-use crate::ErrorKey;
 
-pub trait ECScalar: Mul<Output = Self> + Add<Output = Self> + Sized {
-    type SecretKey;
+/// Elliptic curve implementation
+///
+/// Refers to according implementation of [ECPoint] and [ECScalar].
+pub trait Curve: PartialEq + Clone + fmt::Debug {
+    type Point: ECPoint<Scalar = Self::Scalar>;
+    type Scalar: ECScalar;
+
+    /// Canonical name for this curve
+    const CURVE_NAME: &'static str;
+}
+
+/// Scalar value modulus [group order](Self::group_order)
+///
+/// ## Note
+/// This is a low-level trait, you should not use it directly. See wrappers [Point], [PointZ],
+/// [Scalar], [ScalarZ].
+///
+/// [Point]: super::wrappers::Point
+/// [PointZ]: super::wrappers::PointZ
+/// [Scalar]: super::wrappers::Scalar
+/// [ScalarZ]: super::wrappers::ScalarZ
+///
+/// Trait exposes various methods to manipulate scalars. Scalar can be zero. Scalar must zeroize its
+/// value on drop.
+pub trait ECScalar: Clone + PartialEq + fmt::Debug + 'static {
+    /// Underlying scalar type that can be retrieved in case of missing methods in this trait
+    type Underlying;
 
-    fn new_random() -> Self;
+    /// Samples a random scalar
+    fn random() -> Self;
+
+    /// Constructs a zero scalar
     fn zero() -> Self;
-    fn get_element(&self) -> Self::SecretKey;
-    fn set_element(&mut self, element: Self::SecretKey);
-    fn from(n: &BigInt) -> Self;
-    fn to_big_int(&self) -> BigInt;
-    fn q() -> BigInt;
-    fn add(&self, other: &Self::SecretKey) -> Self;
-    fn mul(&self, other: &Self::SecretKey) -> Self;
-    fn sub(&self, other: &Self::SecretKey) -> Self;
-    fn invert(&self) -> Self;
+    /// Checks if the scalar equals to zero
+    fn is_zero(&self) -> bool;
+
+    /// Constructs a scalar `n % curve_order`
+    fn from_bigint(n: &BigInt) -> Self;
+    /// Converts a scalar to BigInt
+    fn to_bigint(&self) -> BigInt;
```

Comment:
Can we just bound this be `From<BigInt>` and `Into<BigInt>` instead?

Reply:
Should we? `from_bigint` has an important note that `n` must be taken by modulo `group_order`. It might be not obvious from trait bound